Address Mark Nottingham’s Last Call review feedback (#55)
linuxwolf/bmeck-ids: javascript-mjs/draft.md

This separation means that (in the absence of additional information) there are two possible interpretations for any given ECMAScript source text. The TC39 standards body for ECMAScript has determined that media types are outside of their scope of work {{TC39-MIME-ISSUE}}.

> It is not possible to fully determine if a source text of ECMAScript is meant to be parsed using the Module or Script grammar goals based upon content alone. Therefore, scripting environments use out-of-band information in order to determine what goal should be used. Some scripting environments have chosen to adopt the file extension of .mjs for this purpose. The media types in this document allow such out-of-band information to be conveyed.

Reviewer: do not allow, right?

Author: Hmm, no — what do you have in mind? Mark explicitly suggested this phrasing.

The idea is that “officially” registering the extension and mapping it to a MIME type formalizes the existing practice in e.g. Node.js.

Reviewer: `text/javascript` cannot convey Module or Script. `.mjs` in Node.js can, but there's no MIME type equivalent.

I think you or @mnot might have misunderstood something. Or perhaps something changed I'm not aware of?

Author: > `text/javascript` cannot convey Module or Script. `.mjs` in Node.js can, but there's no MIME type equivalent.

Indeed, and unfortunately it’s too late to change that.

> I think you or @mnot might have misunderstood something. Or perhaps something changed I'm not aware of?

Nothing (major) changed since you last reviewed the doc — w.r.t. modules, it still only describes web reality and doesn’t invent anything new. The doc officially introduces the `.mjs` extension and maps it to `text/javascript`. It doesn’t affect browsers.

Standardization of the `.mjs` extension is enough to “allow such out-of-band information to be conveyed”.

Reviewer: Thanks. So in contrast to the suggested text, media types emphatically do not allow that information to be conveyed. Instead you need `<script type=module>` or equivalent on the web.

Mark, there was a `text/javascript;goal=module` proposal at some point, but it didn't gain traction. I think at this point we better clarify that media types cannot be used for this as that's the reality.

Author: The media type doesn't, but the extension does enable this for non-browser environments (as already used in practice in Node.js, `d8`). The new extension is also part of this doc.

Reviewer: So why did you write:

> The media types in this document allow such out-of-band information to be conveyed.

?

Reviewer 2: The previous sentences imply that `.mjs` is used as out-of-band information (i.e., that not contained in the media type itself) to make this decision. The suggestion was to explicitly allow that to happen, because the current wording leaves the reader wondering whether it's a good thing or not.

Reviewer: Oh, I think I get the intent now. I was reading that sentence as if the media types provide that capability themselves as well. How about:

```suggestion
It is not possible to fully determine if a source text of ECMAScript is meant to be parsed using the Module or Script grammar goals based upon content or media type alone. Therefore, as allowed by the media types in this document, scripting environments use out-of-band information in order to determine what goal should be used. Some scripting environments have chosen to adopt the file extension of .mjs for this purpose.
```

Reviewer 2: WFM

Author: Sounds good, thanks!

How implementations determine the character encoding scheme can be subject to processing rules that are out of the scope of this document.  For example, transport protocols can require that a specific character encoding scheme is to be assumed if the optional charset parameter is not specified, or they can require that the charset parameter is used in certain cases.  Such requirements are not considered part of this document.

> Implementations that support binary source text MUST support binary source text encoded using the UTF-8 {{RFC3629}} character encoding scheme.  Module goal sources MUST be encoded as UTF-8, all other encodings will fail.  Source goal sources SHOULD be encoded as UTF-8; other character encoding schemes MAY be supported, but are discouraged.  Whether U+FEFF is processed as a Byte Order Mark (BOM) signature or not depends on the host environment, and is not defined by this document.

Author: @mnot Does this clarify things?

Reviewer 2: That's fine; IIRC there are a few other instances of this wording...

Reviewer: FWIW, the paragraph before has this.

And Security Considerations has "considered out of the scope of this document" three times, which I suppose could use the same wording as well.

Author: Done in latest version. Thanks everyone!
